src/main/java/org/ice4j/ice/ConnectivityCheckClient.java

```diff
@@ -907,118 +1009,19 @@ private long getNextWaitInterval()
-                        {
-                            wait(waitFor);
-                        }
-                        catch (InterruptedException e)
-                        {
-                            logger.log(Level.FINER, "PaceMaker got interrupted",
-                                    e);
-                        }
-
-                        if (!running)
-                            break;
-                    }
-
-                    CandidatePair pairToCheck = checkList.popTriggeredCheck();
-
-                    //if there are no triggered checks, go for an ordinary one.
-                    if (pairToCheck == null)
-                        pairToCheck = checkList.getNextOrdinaryPairToCheck();
-
-                    if (pairToCheck != null)
-                    {
-                        /*
-                         * Since we suspect that it is possible to
-                         * startCheckForPair, processSuccessResponse and only
-                         * then setStateInProgress, we'll synchronize. The
-                         * synchronization root is the one of the
-                         * CandidatePair#setState method.
-                         */
-                        synchronized (pairToCheck)
-                        {
-                            TransactionID transactionID
-                                = startCheckForPair(pairToCheck);
-
-                            if (transactionID == null)
-                            {
-                                logger.info(
-                                        "Pair failed: "
-                                            + pairToCheck.toShortString());
-                                pairToCheck.setStateFailed();
-                            }
-                            else
-                                pairToCheck.setStateInProgress(transactionID);
-                        }
-                    }
-                    else
-                    {
-                        /*
-                         * We are done sending checks for this list. We'll set
-                         * its final state in either the processResponse(),
-                         * processTimeout() or processFailure() method.
-                         */
-                        logger.finest("will skip a check beat.");
-                        checkList.fireEndOfOrdinaryChecks();
-                    }
-                }
-            }
-            finally
-            {
-                synchronized (paceMakers)
-                {
-                    synchronized (this)
-                    {
-                        paceMakers.remove(this);
-                    }
-                }
-            }
-        }
     }
 
     /**
      * Stops and removes all <tt>PaceMaker</tt>s.
      */
     public void stop()
     {
-        synchronized (paceMakers)
+        Iterator<PaceMaker> paceMakersIter = paceMakers.iterator();
+        while(paceMakersIter.hasNext())
```

Comment:
if i'm reading the [docs](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentLinkedQueue.html) correctly, i think we could run into an issue here.  the iterator will be valid (no concurrent modification issues), but, the view is only consistent with when the iterator is created, so if a pacemaker was added after we created the iterator here we won't see it and shut it down.